torch/_inductor/decomposition.py

```diff
@@ -250,10 +250,12 @@ def mm(self, input2):
 
 @register_decomposition([aten.cat.default])
 def cat(tensors, dim=0):
+    from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
+
     def non_empty_tensor(x):
         # special case for cat'ing with an empty tensor -
         # just drop the 'empty' inputs so they don't confuse the logic below.
-        return len(x.shape) > 1 or x.shape[0] > 0
+        return len(x.shape) > 1 or guard_size_oblivious(x.shape[0] > 0)
```

Comment:
I don't this is guard oblivious - if this is shape 0 during runtime and we ignore and it's later shape 1 then we will exclude it and get wrong results. I think this should add runtime deferred asserts

Reply:
Cannot add a deferred runtime assert, there is no error condition on either side of the condition. If the user wants to claim that this shape can never be zero, they need to say so explicitly.

So if this is not size oblivious, should we just prevent this optimization from happening when we don't backed SymInts? I'll look more carefully.

And hooray for size oblivious explicitness paying off!

Reply:
If it's an unbacked symint we should just reshape it to be an empty tensor with the otherwise correct dimensions

Reply:
E.g. 

`Cat (([2, 2], [3, 2], [0]), dim=0) -> Cat(([2, 2], [3, 2], [0, 2]), dim=0)`

Reply:
After thinking carefully about this, I think `guard_size_oblivious` here is correct.

The worrisome case is `cat([(2, 2), (u0,)], dim=0)`. Specifically, `cat([(2, 2), (u0, 2)], dim=0)` is NOT problematic, because it doesn't matter if u0 is zero or not, we are always going to treat it uniformly in both cases (in particular we don't *need* to drop it from cat arguments so that split-cat optimization is right, since split-cat will give us a perfectly reasonable result assuming that u0 is non-zero.)

In the worrisome case, when we guard_size_oblivious here, we will assume that u0 is non-zero. This means that this expression overall will have actually failed before getting here, because you can't cat tensors of different dimensions together. This means that the *only* way this could have passed is if we added a deferred runtime assert that `u0 == 0`, which... we could? But TBH, I don't see any reason to do this.

Reply:
You're good - I was misremembering the details of `guard_size_oblivious` and took it to mean that we would use the size and not guard on it.. But we assume >=2 if size like and unbacked